go-controller/pkg/clustermanager/network_cluster_controller.go

```diff
@@ -0,0 +1,499 @@
+	subnet, err := houtil.ParseHybridOverlayHostSubnet(node)
+	if err != nil {
+		// Log the error and try to allocate new subnets
+		klog.Infof("Failed to get node %s hybrid overlay subnet annotation: %v", node.Name, err)
+	} else if subnet != nil {
+		existingSubnets = []*net.IPNet{subnet}
+	}
+
+	// Allocate a new host subnet for this node
+	// FIXME: hybrid overlay is only IPv4 for now due to limitations on the Windows side
+	hostSubnets, allocatedSubnets, err := ncc.hybridOverlaySubnetAllocator.AllocateNodeSubnets(node.Name, existingSubnets, true, false)
+	if err != nil {
+		return nil, fmt.Errorf("error allocating hybrid overlay HostSubnet for node %s: %v", node.Name, err)
+	}
+
+	if err := annotator.Set(hotypes.HybridOverlayNodeSubnet, hostSubnets[0].String()); err != nil {
+		if e := ncc.hybridOverlaySubnetAllocator.ReleaseNodeSubnets(node.Name, allocatedSubnets...); e != nil {
+			klog.Errorf("Failed to release hybrid over subnet for the node %s from the allocator : %w", node.Name, e)
+		}
+		return nil, fmt.Errorf("error setting hybrid overlay host subnet: %w", err)
+	}
+
+	return hostSubnets[0], nil
+}
+
+func (ncc *networkClusterController) releaseHybridOverlayNodeSubnet(nodeName string) {
+	ncc.hybridOverlaySubnetAllocator.ReleaseAllNodeSubnets(nodeName)
+	klog.Infof("Deleted hybrid overlay HostSubnets for node %s", nodeName)
+}
+
+// handleAddUpdateNodeEvent handles the add or update node event
+func (ncc *networkClusterController) handleAddUpdateNodeEvent(node *corev1.Node) error {
+	if util.NoHostSubnet(node) {
+		if ncc.enableHybridOverlaySubnetAllocator && houtil.IsHybridOverlayNode(node) {
+			annotator := kube.NewNodeAnnotator(ncc.kube, node.Name)
+			allocatedSubnet, err := ncc.hybridOverlayNodeEnsureSubnet(node, annotator)
+			if err != nil {
+				return fmt.Errorf("failed to update node %s hybrid overlay subnet annotation: %v", node.Name, err)
+			}
+			if err := annotator.Run(); err != nil {
+				// Release allocated subnet if any errors occurred
+				if allocatedSubnet != nil {
+					ncc.releaseHybridOverlayNodeSubnet(node.Name)
+				}
+				return fmt.Errorf("failed to set hybrid overlay annotations for node %s: %v", node.Name, err)
+			}
+		}
+		return nil
+	}
+
+	return ncc.syncNodeClusterSubnet(node)
+}
+
+func (ncc *networkClusterController) syncNodeClusterSubnet(node *corev1.Node) error {
+	ncc.clusterSubnetAllocator.Lock()
+	defer ncc.clusterSubnetAllocator.Unlock()
+
+	existingSubnets, err := util.ParseNodeHostSubnetAnnotation(node, ncc.networkName)
+	if err != nil && !util.IsAnnotationNotSetError(err) {
+		// Log the error and try to allocate new subnets
+		klog.Warning("Failed to get node %s host subnets annotations for network %s : %v", node.Name, ncc.networkName, err)
+	}
+
+	hostSubnets, allocatedSubnets, err := ncc.clusterSubnetAllocator.AllocateNodeSubnets(node.Name, existingSubnets, config.IPv4Mode, config.IPv6Mode)
+	if err != nil {
+		return err
+	}
+
+	if len(allocatedSubnets) == 0 {
+		return nil
+	}
```

Comment:
Because if allocatedsubnets len is 0, then we don't have to do anything as no new subnets were allocated.

Otherwise we would end up calling updateNodeAnnotationWithRetry unnecessarily.